File: 07.FamilyTree/code/family.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Iterator;

public class family {

	public static void main(String[] args) throws IOException {
		String fname = "family.inp";
		String outfname = "family.out";
		FileReader fr = new FileReader(fname);
		FileSet(fr, outfname);
	}
	
	public static void FileSet(FileReader fr, String outfname) throws IOException {
		BufferedReader br_f = new BufferedReader(fr);
		BufferedOutputStream bs = null;
		bs = new BufferedOutputStream(new FileOutputStream(outfname));
		
		int arrayListSize = Integer.parseInt(br_f.readLine());
		
		String str = "";
		
		for (int i = 0; i < arrayListSize; i++) {
			Map<Integer, ArrayList<Integer>> map = new HashMap<>();
			
			int dataNum = Integer.parseInt(br_f.readLine());

			ArrayList<Integer> childList = new ArrayList<>();
			ArrayList<Integer> rootKeyList = new ArrayList<>();
			for (int j = 0; j < dataNum-1; j++) {
				
				String[] strArr = br_f.readLine().split(" "); // 3 5
				Integer parent = Integer.valueOf(strArr[0]);
				Integer child = Integer.parseInt(strArr[1]);
				
				if (map.get(parent) == null) {
					map.put(parent, new ArrayList<>(Arrays.asList(child)));
				} else {
					ArrayList<Integer> newChildList = new ArrayList<>(map.get(parent));
					newChildList.add(child);
					map.put(parent, newChildList);
				}
				childList.add(child);
				rootKeyList.add(parent);
			}
			rootKeyList.removeAll(childList);
			int root = rootKeyList.get(0).intValue();
			ArrayList<Integer> keyList = new ArrayList(Arrays.asList(root));
					
			int depth = 0;
			while(true) {
				ArrayList<Integer> childArray = new ArrayList<>();
				int keyLen = keyList.size();
				for (int t=0; t< keyLen ; t++) {
					ArrayList<Integer> nowChildList = map.get(keyList.get(t));
					
					if (nowChildList != null) {
						int nowLen = nowChildList.size();
						for (int k=0; k< nowLen ; k++) {
							childArray.add(nowChildList.get(k));
						}
					}
				}
				keyList = childArray;
				
				depth++;
				if (keyList.size() == 0) break;
			}
			str += depth + "\n";

		}
		bs.write(str.getBytes());

		
		br_f.close();
		bs.close();
	}	
}

File: 07.FamilyTree/code/family.py

rfile = open("6.inp", "r")
wfile = open("family.out", "w")

testCount = int(rfile.readline().strip())

#print testCount
import time
print (time.localtime())
for treeLine in range(testCount) :
    tree = {}
    childList = []
    keyList = []

    lineCount = int(rfile.readline().strip())
    #print(lineCount)
    for i in range(lineCount-1) :
        (parent, child) = rfile.readline().strip().split(" ")
        # print line
        # parent = line[0]
        # child = line[1]
        keyList.append(parent)
        childList.append(child)

        if parent in tree :
            temp = tree[parent]
            temp.append(child)
            tree[parent] = temp
        else :
            tree[parent] = [child]

    #print tree 

    parentArray = list(set(keyList) - set(childList))
    depth = 0
    #print root

    #print(parentArray)
    # break
    while True :
        nowChild = []
        for key in parentArray :
            if key in tree :
                nowChild += tree[key]

        #print nowChild
        
        
        parentArray = nowChild
        
        depth += 1
        # print(depth)
        if len(parentArray) == 0 : break


    # print depth
    wfile.write("%d\n" % depth)
    # break
print (time.localtime())
rfile.close()
wfile.close()


